### PR TITLE
Added SmartGPT workflow (issue #4463)

### LINKED ISSUE #4463: `SmartGPT` workflow
### Feature request

@hwchase17   Can we implement this [**SmartGPT** workflow](https://youtu.be/wVzuvf9D9BU)? 
Probably, it is also implemented but I didn't find it.
Thi method looks like something simple but very effective.

### Motivation

Improving the quality of the prompts and the resulting generation quality.

### Your contribution

I can try to implement it but need direction.

### PR DESCRIPTION
**Pull request by UmerHA**
# Added SmartGPT workflow by providing SmartLLM wrapper around LLMs
Edit:
As @hwchase17 suggested, this should be a chain, not an LLM. I have adapted the PR.

It is used like this:
```
from langchain.prompts import PromptTemplate
from langchain.chains import SmartLLMChain
from langchain.chat_models import ChatOpenAI

hard_question = "I have a 12 liter jug and a 6 liter jug. I want to measure 6 liters. How do I do it?"
hard_question_prompt = PromptTemplate.from_template(hard_question)

llm = ChatOpenAI(model_name="gpt-4")
prompt = PromptTemplate.from_template(hard_question)
chain = SmartLLMChain(llm=llm, prompt=prompt, verbose=True)

chain.run({})
```

Original text: 
Added SmartLLM wrapper around LLMs to allow for SmartGPT workflow (as in https://youtu.be/wVzuvf9D9BU). SmartLLM can be used wherever LLM can be used. E.g:

```
smart_llm = SmartLLM(llm=OpenAI())
smart_llm("What would be a good company name for a company that makes colorful socks?")
```
or
```
smart_llm = SmartLLM(llm=OpenAI())
prompt = PromptTemplate(
    input_variables=["product"],
    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=smart_llm, prompt=prompt)
chain.run("colorful socks")
```

SmartGPT consists of 3 steps:

1. Ideate - generate n possible solutions ("ideas") to user prompt
2. Critique - find flaws in every idea & select best one
3. Resolve - improve upon best idea & return it

Fixes #4463

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

- @hwchase17
- @agola11

Twitter: [@UmerHAdil](https://twitter.com/@UmerHAdil) | Discord: RicChilligerDude#7589

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
i think this is probably more of chain than an LLM (in the same way the constitutional AI chain exists

**Comment by UmerHA** [comment]
Agreed - will change

**Comment by UmerHA** [comment]
Requires https://github.com/hwchase17/langchain/pull/5030 to be merged. Will re-request review, when it's merged.

File: langchain/schema.py [hunk 1]
```diff
@@ -128,6 +128,7 @@ def messages_to_dict(messages: List[BaseMessage]) -> List[dict]:
 
 def _message_from_dict(message: dict) -> BaseMessage:
     _type = message["type"]
+    print(f"schema._message_from_dict: Parsing {message}")
```
Comment (hwchase17) [inline langchain/schema.py:131]: stray print

File: langchain/chat_models/fake.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+"""Fake ChatModel for testing purposes."""
```
Comment (hwchase17) [inline langchain/chat_models/fake.py:1]: nice love this

File: langchain/chains/smart_llm/base.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        if stage == "ideation":
+            return role_strings
+        role_strings.extend(
+            [
+                *[
+                    ("ai", "Idea " + str(i + 1) + ": {idea_" + str(i + 1) + "}")
+                    for i in range(self.n_ideas)
+                ],
+                (
+                    "human",
```
Comment (hwchase17) [inline langchain/chains/smart_llm/base.py:203]: from your error message, i would actually use `from_strings` with something like `(HumanMessage, "fooo....")`

Comment (UmerHA) [reply langchain/chains/smart_llm/base.py]: Done

File: langchain/schema.py (hunk 1 above)
Comment (UmerHA) [reply langchain/schema.py]: removed

File: docs/modules/chains/examples/smart_llm.ipynb [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "811ea8e1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hard_question = \"I have a 12 liter jug and a 6 liter jug. I want to measure 6 liters. How do I do it?\"\n",
```
Comment (leo-gan) [inline docs/modules/chains/examples/smart_llm.ipynb:74]: Sorry, I didn't get it. 
To measure 6 liter you have to fill in the 6 liter jug. That's all.
The answer definitely didn't provide a good answer.
Probably, the question can be like the puzzle for the "Die hard" move:
Given a 5-gallon jug and a 3-gallon jug. Please, fill one with exactly 4 gallons of water.

Comment (UmerHA) [reply docs/modules/chains/examples/smart_llm.ipynb]: Hey, I disagree :) This is the example shown in the original SmartGPT video (https://youtu.be/wVzuvf9D9BU?t=384). The point is none of the original ideas work at all (even with GPT4!), only with the SmartGPT workflow do we get a correct answer.

But I agree this is confusing without context. I added more explanation.

**Comment by UmerHA** [comment]
Hey @hwchase17 any more comments?

**Comment by batmanscode** [comment]
Nice work! 😃

**Comment by UmerHA** [comment]
Hi @hinthornw! Also pinging you here, as you're the maintainer for Agents :) Any feedback?

**Comment by UmerHA** [comment]
> @UmerHA do you think this could be used for a pandas/xorbits agent? To break down complex questions before answering

Sure, why not. This is a general approach and should improve any task (at the cost of more token usage).

**Comment by batmanscode** [comment]
> > @UmerHA do you think this could be used for a pandas/xorbits agent? To break down complex questions before answering
> 
> Sure, why not. This is a general approach and should improve any task (at the cost of more token usage).

Interesting thanks. Must look into using this as an agent with tools!

**Comment by baskaryan** [comment]
apologies for the delay in reviewing, this is very cool! thank you @UmerHA!!

**Pull request merged**